composer/utils/collect_env.py

```diff
@@ -0,0 +1,391 @@
+    EXCEPTION_MSG = f"Bugs can be reported at: {COMPOSER_OPEN_ISSUE_URL}\n" + \
+                    f"Please include details on how to reproduce the issue and attach the following environment report:\n"
+
+    # Don't print exception report for KeyboardInterrupt
+    if not issubclass(exc_type, KeyboardInterrupt):
+        if issubclass(exc_type, AssertionError):
+            EXCEPTION_SEV_MSG = f"Smells like a Composer bug.\n"
+        elif issubclass(exc_type, RuntimeError):
+            EXCEPTION_SEV_MSG = f"This could be due to user error but is most likely a Composer bug.\n"
+        elif issubclass(exc_type, ValueError) or issubclass(exc_type, TypeError):
+            EXCEPTION_SEV_MSG = f"This was most likely due to user error but please submit a bug report if you suspect a Composer issue.\n"
+        else:
+            EXCEPTION_SEV_MSG = f"If you would like support debugging, submit a bug report or reach out to us on our community channels.\n"
+
+        print("\n-------------------\n" + EXCEPTION_SEV_MSG + EXCEPTION_MSG, file=sys.stderr)
+        print_env(sys.stderr)
+
+
+# Excepthook wrapper, wraps default excepthook and prints env info
+def _custom_exception_handler(type, value, tb) -> None:
+    """Custom exception wrapper for sys.excepthook."""
+    _exc_report(exc_type=type)
+    assert _orig_excepthook
+    _orig_excepthook(type, value, tb)
+
+
+# Custom exception handler for IPython notebooks
+def _nb_custom_exception_handler(self, type, value, tb, tb_offset=None):
+    """Custom exception handler for IPython."""
+    _exc_report(exc_type=type)
+    self.showtraceback((type, value, tb), tb_offset=tb_offset)  # standard IPython's printout
+
+
+# Public function to register excethook wrapper
+def configure_excepthook() -> None:
+    """Collect and print system information when :func:`sys.excepthook` is called.
+
+    This function checks if the user is running from an IPython session and sets up the custom
+    exception handler accordingly.
+
+    To override the default :func:`sys.excepthook` with the custom except hook:
+
+    .. testsetup:: composer.utils.collect_env.configure_excepthook
+
+        import sys
+        from composer.utils.collect_env import configure_excepthook
+
+        sys.excepthook = sys.__excepthook__
+
+    .. doctest:: composer.utils.collect_env.configure_excepthook
+
+        >>> configure_excepthook()
+        >>> sys.excepthook 
+        <function _custom_exception_handler at ...>
+    """
+
+    # Custom exceptions work differntly in notebooks
+    if IPYTHON_AVAILABLE:
+        # Set custom handler on Exception base class to apply to all exceptions
+        nb.set_custom_exc((Exception,), _nb_custom_exception_handler)
+    else:
+        # Save original excepthook and override
+        global _orig_excepthook
+        _orig_excepthook = sys.excepthook
+        sys.excepthook = _custom_exception_handler
+
+
+# Public function to restore original excepthook
+def restore_excepthook() -> None:
```

Comment:
If another function modified `sys.excepthook` after calling `configure_excepthook`, then when it is restored, the other excepthook would also be removed, which is probably undesired behavior. For example:

```python
configure_excepthook()
sys.excepthook = thidparty_except_hook_handler(sys.excepthook)
restore_excepthook()  # would also remove thirdparty_except_hook_handler, since we are restoring the original excepthook
```

Instead, could the `restore_excepthook` function make the `_custom_exception_handler` or `_nb_custom_exception_handler` no-ops (but still leave it in the stack?) That would preserve the ability of `thirdparty_except_hook_handler` to still fire, even though our excepthook would now be a noop.